storage: reactive gossiping of store capacity based on rate of change

spencerkimball opened the pull request:
Stores periodically gossip their store descriptors, which contain capacity
information used to rebalance replicas and leases. This was previously
being done on a fixed, 5s interval. For small clusters with few ranges,
this would often lead to degenerate rebalancing scenarios because 5s
wouldn't be fast enough to prevent rebalancing based on stale information.

This change uses a fairly coarse-grained measure of changes, counting any
update to a lease or addition/subtraction of a replica against a countdown,
measured as a fraction of total replicas and leaseholders. If the countdown
completes before the normal store gossip interval, the store gossips early.
The default store gossip interval has also been relaxed to 60s.

Fixes #9242

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/11662)
<!-- Reviewable:end -->

spencerkimball committed:
storage: reactive gossiping of store capacity based on rate of change

Stores periodically gossip their store descriptors, which contain capacity
information used to rebalance replicas and leases. This was previously
being done on a fixed, 5s interval. For small clusters with few ranges,
this would often lead to degenerate rebalancing scenarios because 5s
wouldn't be fast enough to prevent rebalancing based on stale information.

This change uses a fairly coarse-grained measure of changes, counting any
update to a lease or addition/subtraction of a replica against a countdown,
measured as a fraction of total replicas and leaseholders. If the countdown
completes before the normal store gossip interval, the store gossips early.
The default store gossip interval has also been relaxed to 60s.

spencerkimball commented:
Fixes #9242

tamird commented:
Please add the "fixes ..." to the commit message.

tamird commented:
Can you add a test, however basic, of this functionality?

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/server/updates.go, line 214 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfjWZTW3obj_2X2EUa:-KXfjWZTW3obj_2X2EUb:b-d3kv9o) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/server/updates.go#L214)):*
> ```Go
> 	if l := len(r.Details); l > 0 {
> 		v := r.Details[l-1]
> 		log.Infof(ctx, "%d new version(s) are available; latest is %s (%s)", l, v.Version, v.Details)
> ```

this doesn't seem related to the rest of the change being made here. what's worse, it is not safe to do this in the general case - some upgrades may require stepping through intervening versions; we cannot blindly advise operators to upgrade to the latest version.

this will need to be addressed on the server side, I suspect.

---

*[pkg/storage/replica_proposal.go, line 389 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfkw6jSxtwCrepA4c_:-KXfkw6jSxtwCrepA4ca:b-gemhlu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/replica_proposal.go#L389)):*
> ```Go
> 
> 	// Notify the store that a lease change occurred and it may need to
> 	// gossip updated capacity.
> ```

s/updated capacity/the updated store descriptor/

---

*[pkg/storage/store.go, line 371 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfkdErfsyqL3wtPZpF:-KXfkdErfsyqL3wtPZpG:b-osc9jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L371)):*
> ```Go
> 
> 	// gossipCapacityCountdown is the number of changes to range count
> 	// or lease count allowed before re-gossiping store descriptor
> ```

missing article; "the" store descriptor

---

*[pkg/storage/store.go, line 372 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfkjM1bzezPUTNY--j:-KXfkjM1bzezPUTNY--k:b-8l0xf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L372)):*
> ```Go
> 	// gossipCapacityCountdown is the number of changes to range count
> 	// or lease count allowed before re-gossiping store descriptor
> 	// earlier than the normal periodic gossip interval. Updated
> ```

s/earlier than the normal periodic gossip interval/(than the normal periodic gossip interval

---

*[pkg/storage/store.go, line 1358 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXflG0e035BDefa1Um3:-KXflG0e035BDefa1Um4:bhwl5a3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L1358)):*
> ```Go
> 	rangeCountdown := float64(storeDesc.Capacity.RangeCount) * gossipWhenCapacityDeltaExceedsFraction
> 	leaseCountdown := float64(storeDesc.Capacity.LeaseCount) * gossipWhenCapacityDeltaExceedsFraction
> 	// Store the minimum value of the two, maximum 1.
> ```

looks like you're storing the maximum of the two.

also looks like you could find the minimum by operating on integers and make `gossipWhenCapacityDeltaExceedsFraction` a divisor (i.e. `100`) and avoid float operations altogether.

---

*[pkg/storage/store.go, line 1397 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfkHCtw6wF3pTMBOcX:-KXfkHCtw6wF3pTMBOcY:b6rclk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L1397)):*
> ```Go
> }
> 
> // In case the capacity values have changed enough since the last time
> ```

what is this comment documenting? doc comments should be written such that they are useful to users of the method, and this doesn't seem to be.

---

*[pkg/storage/store.go, line 1399 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfk9mG_Xi5X0xYDbNq:-KXfk9mG_Xi5X0xYDbNr:b-xa573i) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L1399)):*
> ```Go
> // In case the capacity values have changed enough since the last time
> // they were gossiped for the store, re-gossip them immediately.
> func (s *Store) maybeGossipOnCapacityChange() {
> ```

can you plumb a context? the caller or grandcaller of this method seems to typically have a context.

---

*[pkg/storage/store.go, line 1405 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfkYuIvirKtej5dDPD:-KXfkYuIvirKtej5dDPE:b-udytfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L1405)):*
> ```Go
> 	go func() {
> 		if err := s.GossipStore(context.TODO()); err != nil {
> 			log.Warningf(context.TODO(), "error doing initial gossiping: %s", err)
> ```

initial?

---

*[pkg/storage/store.go, line 2132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfk-HQ8XeJe9ggVIZU:-KXfk-HQ8XeJe9ggVIZV:b-cwffye) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L2132)):*
> ```Go
> }
> 
> func (s *Store) incReplicaCountLocked() {
> ```

i don't think these methods actually require any locks to be held.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-:-KXfm4xhDmTCcqmSM9p7:bps2ctf)*
<!-- Sent from Reviewable.io -->

petermattis commented:
How often does gossiping occur in practice with this change? 

Note that returning the current store capacity in `SnapshotResponse.StoreCapacity` is supposed to more quickly spread capacity information when lots of rebalancing is occurring. Did that mechanism get broken?

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store.go, line 2132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfk-HQ8XeJe9ggVIZU:-KXfmxmo7jFzYFmAO6I_:b-d0cvpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L2132)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i don't think these methods actually require any locks to be held.
</blockquote></details>

Sort of strange to me that `{inc,dec}ReplicaCountLocked` have the side effect of possibly kicking off a store gossip. Also, if you look closely you'll notice that the number of replicas we gossip is `len(s.mu.replicas)`, not `s.metrics.ReplicaCount`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-:-KXfnJKnWJTIqb0GbswW:bvcg8ib)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
@petermattis, in a small cluster it gossips mostly due to lease updates; so that ends up being roughly the same as the 5s interval, only it gossips immediately when things change instead of exactly once every 5s.

I didn't know about `SnapshotResponse.StoreCapacity`. It's certainly active, but doesn't actually work to help with the rebalancing thrashing because it isn't sending updated range and leaseholder counts, which are what we actually rebalance on. When I fix that oversight, the snapshot response capacity update still has the following problems:
  - It takes 29 total snapshots to rebalance a fresh five node cluster (versus 16 for reactive gossip)
  - The leaseholders do not seem to rebalance as expected
  - Only some nodes have accurate stats, so you end up requiring periodic gossip to be fairly fast regardless (i.e. we will not be able to move from 5s to something more reasonable for large clusters, like 60s).

My feeling is we should remove `SnapshotResponse.StoreCapacity`. It's not implemented correctly at the moment and even if it were, it still feels like it will at most rise only to the level of a half measure.

@tamird added a unittest.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/server/updates.go, line 214 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfjWZTW3obj_2X2EUa:-KXg6AE3VzhfSE5fmMyl:bsfapzo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/server/updates.go#L214)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this doesn't seem related to the rest of the change being made here. what's worse, it is not safe to do this in the general case - some upgrades may require stepping through intervening versions; we cannot blindly advise operators to upgrade to the latest version.

this will need to be addressed on the server side, I suspect.
</blockquote></details>

I'll just back this out. It struck me as over the top how many of these things we're printing out on startup. Perhaps we should be doing something more clever.

---

*[pkg/storage/replica_proposal.go, line 389 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfkw6jSxtwCrepA4c_:-KXg6RAXDdDpHvYAOEEU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/replica_proposal.go#L389)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/updated capacity/the updated store descriptor/
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 371 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfkdErfsyqL3wtPZpF:-KXg6dDRtFcnZohOjycW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L371)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

missing article; "the" store descriptor
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 372 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfkjM1bzezPUTNY--j:-KXg6lxFKz8iDCLnAXRA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L372)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/earlier than the normal periodic gossip interval/(than the normal periodic gossip interval
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 1358 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXflG0e035BDefa1Um3:-KXg6w5jHGGU0RbyfVFO:b-auwt0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L1358)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

looks like you're storing the maximum of the two.

also looks like you could find the minimum by operating on integers and make `gossipWhenCapacityDeltaExceedsFraction` a divisor (i.e. `100`) and avoid float operations altogether.
</blockquote></details>

Oops on max...fixed.

What's the problem with floating point? In any case, I think it's easier to understand this way.

---

*[pkg/storage/store.go, line 1397 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfkHCtw6wF3pTMBOcX:-KXg7M_WaQkJHUPHKSfi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L1397)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what is this comment documenting? doc comments should be written such that they are useful to users of the method, and this doesn't seem to be.
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 1399 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfk9mG_Xi5X0xYDbNq:-KXgN19mGU3we7VijIZM:bzbsbrx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L1399)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you plumb a context? the caller or grandcaller of this method seems to typically have a context.
</blockquote></details>

Since we're kicking off a child goroutine to do the processing, I'm instead using a store-annotated `context.Background()`.

---

*[pkg/storage/store.go, line 1405 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfkYuIvirKtej5dDPD:-KXg7Z3WYlQ4R4qtCBi9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L1405)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

initial?
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 2132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfk-HQ8XeJe9ggVIZU:-KXg7a5SjKPjIivehTiQ:b-fvxep9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L2132)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Sort of strange to me that `{inc,dec}ReplicaCountLocked` have the side effect of possibly kicking off a store gossip. Also, if you look closely you'll notice that the number of replicas we gossip is `len(s.mu.replicas)`, not `s.metrics.ReplicaCount`. 
</blockquote></details>

I'm aware we don't gossip `s.metrics.ReplicaCount`, though this metric will be exactly equal to `len(s.mu.replicas)`. It just seemed better to break this off into a method instead of adding a call to `maybeGossipOnCapacityChange` everywhere we add or remove something from `s.mu.replicas`. Other ideas?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-:-KXfstr1i6O74fXw5ZTM:bqk5nkv)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> When I fix that oversight, the snapshot response capacity update still has the following problems

When I fix the problem with `Store.Capacity()`, I see only 11 snapshots. This is by running: 

```
~ rm -fr cockroach-data && roachdemo -n 5 -- --logtostderr
...
~ egrep 'applied.*snapshot' cockroach-data/*/logs/*.stderr | wc -l
11
```

I think we should fix `Store.Capacity` in addition to this change. It is a lurking error that it doesn't fill in the fields. In particular, I was testing using this diff:

```diff
diff --git a/pkg/storage/store.go b/pkg/storage/store.go
index e2dd057..02fd5b5 100644
--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -2126,7 +2126,12 @@ func (s *Store) Attrs() roachpb.Attributes {
 // Capacity returns the capacity of the underlying storage engine. Note that
 // this does not include reservations.
 func (s *Store) Capacity() (roachpb.StoreCapacity, error) {
-       return s.engine.Capacity()
+       capacity, err := s.engine.Capacity()
+       if err == nil {
+               capacity.RangeCount = int32(s.ReplicaCount())
+               capacity.LeaseCount = int32(s.LeaseCount())
+       }
+       return capacity, err
 }

 // Registry returns the store registry.
@@ -2155,8 +2160,6 @@ func (s *Store) Descriptor() (*roachpb.StoreDescriptor, error) {
        if err != nil {
                return nil, err
        }
-       capacity.RangeCount = int32(s.ReplicaCount())
-       capacity.LeaseCount = int32(s.LeaseCount())
        // Initialize the store descriptor.
        return &roachpb.StoreDescriptor{
                StoreID:  s.Ident.StoreID,
```

Please hold off on merging this PR until we have #11591 sorted out. 

---

Review status: 1 of 9 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[vendor, line 1 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXn4bA-Q43Ej5gqqQj0:-KXn4bA-Q43Ej5gqqQj1:bd74glx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d67a86696f2ea6259305b887b45ae0cdf06892a/vendor#L1)):*
> ```
> Subproject commit 7f1656b1f8ed0acd6a6f3342355382cdde9cac57
> ```

You probably didn't mean to do this. I think you need to run `git submodule update`.

---

*[pkg/storage/store.go, line 2132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfk-HQ8XeJe9ggVIZU:-KXn4ne-6aM-U3rwRmNU:b-ssw70s) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L2132)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I'm aware we don't gossip `s.metrics.ReplicaCount`, though this metric will be exactly equal to `len(s.mu.replicas)`. It just seemed better to break this off into a method instead of adding a call to `maybeGossipOnCapacityChange` everywhere we add or remove something from `s.mu.replicas`. Other ideas?
</blockquote></details>

I'm happier with the handful of calls to `maybeGossipOnCapacityChange` as you have now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-:-KXn3qs5y5XL83jm5H6N:b-gysiqe)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Well that's probably because you're also still gossiping every `5s`. I had that set to `60s`, which feels a lot more appropriate when you're talking about anything other than an empty cluster. Why should every node go through 10,000 or even 100,000 ranges every 5s to count leaseholders? When that `5s` change was made, it was agreed that it was a temporary evil and I dislike this new normal that it's become a default setting.

petermattis commented:
> Well that's probably because you're also still gossiping every 5s.

When I set the gossip interval to `60s` I only saw 14 snapshots. `Store.LeaseCount` is currently O(num-replicas), but it doesn't need to be.

petermattis commented:
Instead of a precise lease count, we could change `Store.LeaseCount` to return the number of leases for which the store was the last lease holder, possibly including expired leases. That count could be maintained incrementally making `Store.LeaseCount` O(1).

---

Review status: 1 of 9 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-:-KXnB3LBqQE5Oz4-Jinz:b-bgfhqe)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
I'm not quite understanding your proposal re: `LeaseCount`. Are you also suggesting that fixing what's returned in capacity will allow us to reset the default to 60s?

---

Review status: 1 of 8 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[pkg/storage/store.go, line 2132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfk-HQ8XeJe9ggVIZU:-KXnAjtTAMoERPCTzAn3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L2132)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm happier with the handful of calls to `maybeGossipOnCapacityChange` as you have now.
</blockquote></details>

Done.

---

*[vendor, line 1 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXn4bA-Q43Ej5gqqQj0:-KXnBjw5IbT1h1fslsha:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7864203d55d68bead4a141ad68aa3fce51ab3e6b/vendor#L1)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You probably didn't mean to do this. I think you need to run `git submodule update`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-:-KXnAtJFW56N3tVfmXdZ:b50dsww)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Well testing again they're roughly equal, but I get 15 rebalancings with reactive gossip and 18 without fairly consistently. Perhaps it has something to do with you using `roachdemo` and how quickly that starts new nodes. I think that's a wash. What I don't understand about the rebalance response capacity update is what it will do when there are many more nodes. When I go to 6 nodes, I get 15 and 20. When I go to 7 nodes, I get 15 and 23. This mechanism seems flawed.

---

Review status: 1 of 8 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-:-KXnE_9HFrZ_qlD6vc0W:ba5gcnn)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Investigating a little further shows that you can often have a situation where rebalancing re-occurs every minute, on the minute, with 60s store gossip interval in the case where we only send updates to range counts with the snapshot responses. There's just not enough information being shared with that mechanism. This is with five nodes all started within a few seconds of each other.

The snapshot response update amounts to a half measure.

bdarnell commented:
LGTM

---

Reviewed 1 of 4 files at r1, 7 of 9 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_proposal.go, line 390 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXox2-rYo7Pu-z4baEy:-KXox2-spx0lxbnR9AoD:bmizrbs) ([raw file](https://github.com/cockroachdb/cockroach/blob/eff3d2cbc3422b7410c547c58c35ff8c39a720b8/pkg/storage/replica_proposal.go#L390)):*
> ```Go
> 	// Notify the store that a lease change occurred and it may need to
> 	// gossip updated store descriptor (with updated capacity).
> 	r.store.maybeGossipOnCapacityChange()
> ```

This method is called for all lease events, including extensions (which don't change capacity anywhere) and when the lease is transferred between two other replicas (which change capacity on those replicas but not here). This should be guarded by a check for `isLeaseChangingHands` and that either `lease.StoreID` or `prevLease.StoreID` is our store. 

---

*[pkg/storage/store.go, line 1358 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXflG0e035BDefa1Um3:-KXoyZoLujHlETLLQ4sF:b-rwjk8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L1358)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Oops on max...fixed.

What's the problem with floating point? In any case, I think it's easier to understand this way.
</blockquote></details>

Why are these combined into a single value? LeaseCount can never be greater than RangeCount, so it's always the one that will be used here. And then you're counting range add/removal events against a countdown based on lease events, which feels wrong. 

---

*[pkg/storage/store.go, line 91 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXoxohT3-KD3mSsj_NH:-KXoxohUsaw4bin2IUA3:b975gqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/eff3d2cbc3422b7410c547c58c35ff8c39a720b8/pkg/storage/store.go#L91)):*
> ```Go
> 	replicaRequestQueueSize = 100
> 
> 	// defaultGossipWhenCapacityDeltaExceedsFraction indicates the default
> ```

This constant isn't exported so it doesn't need a comment unless there's something more interesting to say. 

---

*[pkg/storage/store.go, line 1413 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXp-CCr96A-xHF0S7WT:-KXp-CCr96A-xHF0S7WU:b-cat97m) ([raw file](https://github.com/cockroachdb/cockroach/blob/eff3d2cbc3422b7410c547c58c35ff8c39a720b8/pkg/storage/store.go#L1413)):*
> ```Go
> 		return
> 	}
> 	go func() {
> ```

We don't use goroutines for other cases when we gossip, even from processRaft goroutines (e.g. maybeGossipSystemConfig from replica_proposal.go). I don't think we need to use a goroutine here. And if we do need a goroutine, it should be managed by the stopper with RunAsyncTask (maybe a limited task just to ensure that these goroutines can't build up?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-:-KXp1C-s9pAS3TkHj-e2:bb6ywmx)*
<!-- Sent from Reviewable.io -->

petermattis commented:
My proposal regarding `LeaseCount` is to make a separate PR to change its O(num-ranges) implementation to an O(1) implementation. Or file an issue.

My other suggestion is that we should fix `Store.Capacity` in this PR. Doing so requires moving 2 lines from `Store.Descriptor` into `Store.Capacity`. I agree that the 5s gossip interval was temporary and moving this interval back to 60s is desirable long term. And I'm fine with this reactive gossiping. I'm not sure that you should remove `SnapshotResult.Capacity` in this change. That feels like turning too many knobs at once. 

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store.go, line 1413 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXp-CCr96A-xHF0S7WT:-KXpQQokwsopmh_MqeJ_:bbk6ske) ([raw file](https://github.com/cockroachdb/cockroach/blob/eff3d2cbc3422b7410c547c58c35ff8c39a720b8/pkg/storage/store.go#L1413)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We don't use goroutines for other cases when we gossip, even from processRaft goroutines (e.g. maybeGossipSystemConfig from replica_proposal.go). I don't think we need to use a goroutine here. And if we do need a goroutine, it should be managed by the stopper with RunAsyncTask (maybe a limited task just to ensure that these goroutines can't build up?)
</blockquote></details>

I think the goroutine is needed to avoid deadlocks as this is called from contexts where `Store.mu` is held.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-:-KXpOu3H6fYLA1fj1Xek:b-bdtfe3)*
<!-- Sent from Reviewable.io -->

tamird commented:
We want this not to go in before
https://github.com/cockroachdb/cockroach/pull/11699 is merged, right?

On Wed, Nov 30, 2016 at 8:31 AM, Peter Mattis <notifications@github.com>
wrote:

> My proposal regarding LeaseCount is to make a separate PR to change its
> O(num-ranges) implementation to an O(1) implementation. Or file an issue.
>
> My other suggestion is that we should fix Store.Capacity in this PR.
> Doing so requires moving 2 lines from Store.Descriptor into Store.Capacity.
> I agree that the 5s gossip interval was temporary and moving this interval
> back to 60s is desirable long term. And I'm fine with this reactive
> gossiping. I'm not sure that you should remove SnapshotResult.Capacity in
> this change. That feels like turning too many knobs at once.
> [image: :lgtm:]
> <https://camo.githubusercontent.com/41b3ec3419116e3b3f84507ad965646ce4ce1f0c/68747470733a2f2f72657669657761626c652e696f2f6c67746d2e706e67>
> ------------------------------
>
> Review status: all files reviewed at latest revision, 12 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *pkg/storage/store.go, line 1413 at r3
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXp-CCr96A-xHF0S7WT:-KXpQQokwsopmh_MqeJ_:bbk6ske>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/eff3d2cbc3422b7410c547c58c35ff8c39a720b8/pkg/storage/store.go#L1413>):*
> *Previously, bdarnell (Ben Darnell) wrote…*
>
> We don't use goroutines for other cases when we gossip, even from
> processRaft goroutines (e.g. maybeGossipSystemConfig from
> replica_proposal.go). I don't think we need to use a goroutine here. And if
> we do need a goroutine, it should be managed by the stopper with
> RunAsyncTask (maybe a limited task just to ensure that these goroutines
> can't build up?)
>
> I think the goroutine is needed to avoid deadlocks as this is called from
> contexts where Store.mu is held.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-:-KXpOu3H6fYLA1fj1Xek:b-bdtfe3>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/11662#issuecomment-263873739>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPPmmqCis1xJMMkHPHaKH5kiFrIRYks5rDXq9gaJpZM4K99ST>
> .
>

petermattis commented:
> We want this not to go in before #11699 is merged, right?

Correct. Let's get #11699 merged and deployed to cyan. And then spend some time writing a robust test for that issue.

tamird commented:
Reviewed 7 of 9 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_split_test.go, line 1749 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXpTiaCPtvJWlzpiuR1:-KXpTiaCPtvJWlzpiuR2:b293akw) ([raw file](https://github.com/cockroachdb/cockroach/blob/eff3d2cbc3422b7410c547c58c35ff8c39a720b8/pkg/storage/client_split_test.go#L1749)):*
> ```Go
> 	<-store.Gossip().Connected
> 
> 	rangeCountCh := make(chan int32, 20)
> ```

why 20? i think this should be unbuffered (with the appropriate safeguards against deadlocking) or else this is risking flakiness under load.

---

*[pkg/storage/client_split_test.go, line 1800 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXpUP5sSn67SHpEmV4m:-KXpUP5sSn67SHpEmV4n:beuawp4) ([raw file](https://github.com/cockroachdb/cockroach/blob/eff3d2cbc3422b7410c547c58c35ff8c39a720b8/pkg/storage/client_split_test.go#L1800)):*
> ```Go
> 	select {
> 	case <-rangeCountCh:
> 		t.Errorf("expected no further gossip")
> ```

have you stressed this test locally? looks like it didn't flake in CI but i'm still worried.

---

*[pkg/storage/replica_proposal.go, line 389 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfkw6jSxtwCrepA4c_:-KXpSTGH6LDUpMaPHdHI:b-j4i5fd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/replica_proposal.go#L389)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

missed the article

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
@petermattis, I made the requested changes to the `Store.Descriptor()` and `Store.Capacity()`, but I'd like to keep the change to remove `SnapshotResult.Capacity`. Even though more things are changing this way, we avoid having a messy situation with two different mechanisms contributing duplicative efforts. While this is all still fresh (and I have relatively high confidence besides), I'd like to see it deployed on its own so in the event it fails, I don't have to only find out about that sometime next week when we get around to deploying the removal of `SnapshotResult.Capacity`.

---

Review status: 5 of 8 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_split_test.go, line 1749 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXpTiaCPtvJWlzpiuR1:-KXpb_XwzQckElOi7kbF:b-e6bs2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/eff3d2cbc3422b7410c547c58c35ff8c39a720b8/pkg/storage/client_split_test.go#L1749)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why 20? i think this should be unbuffered (with the appropriate safeguards against deadlocking) or else this is risking flakiness under load.
</blockquote></details>

It didn't need to be. I had a lot of trouble writing a test for this and this was vestigial. The current test formulation is quite resistant to flakiness.

---

*[pkg/storage/client_split_test.go, line 1800 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXpUP5sSn67SHpEmV4m:-KXpbhWf2ZZHA0Qs4ys_:b-qeargx) ([raw file](https://github.com/cockroachdb/cockroach/blob/eff3d2cbc3422b7410c547c58c35ff8c39a720b8/pkg/storage/client_split_test.go#L1800)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

have you stressed this test locally? looks like it didn't flake in CI but i'm still worried.
</blockquote></details>

Yes, ran it 10,000 times locally.

---

*[pkg/storage/replica_proposal.go, line 389 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXfkw6jSxtwCrepA4c_:-KXpbr_GfdIfVyxA6gJ3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/replica_proposal.go#L389)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

missed the article
</blockquote></details>

Done.

---

*[pkg/storage/replica_proposal.go, line 390 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXox2-rYo7Pu-z4baEy:-KXpcUXLT4-wDJhCXT-Y:b-3dfwst) ([raw file](https://github.com/cockroachdb/cockroach/blob/eff3d2cbc3422b7410c547c58c35ff8c39a720b8/pkg/storage/replica_proposal.go#L390)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This method is called for all lease events, including extensions (which don't change capacity anywhere) and when the lease is transferred between two other replicas (which change capacity on those replicas but not here). This should be guarded by a check for `isLeaseChangingHands` and that either `lease.StoreID` or `prevLease.StoreID` is our store. 
</blockquote></details>

My thought was to future proof by erring on the side of coarse-grained caution. But I agree perhaps it's too coarse-grained, so changed to match your suggestion.

---

*[pkg/storage/store.go, line 1358 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXflG0e035BDefa1Um3:-KXpd76cJWmkKFDUianx:b48fu2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/7512a59908fa654448d862a02a24cd523433b7b3/pkg/storage/store.go#L1358)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why are these combined into a single value? LeaseCount can never be greater than RangeCount, so it's always the one that will be used here. And then you're counting range add/removal events against a countdown based on lease events, which feels wrong. 
</blockquote></details>

I wanted to avoid two values for simplicity. Perhaps this is just more confusing.

We do need both values though, because we're taking the min, not the max between the two. We want to gossip quickly, for example, if a store has just 100 leases and gains a new one, even if it also has a much larger array of 10,000 ranges.

I've split them into two counts.

---

*[pkg/storage/store.go, line 91 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXoxohT3-KD3mSsj_NH:-KXpd-EDpLQomIbVUx5y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eff3d2cbc3422b7410c547c58c35ff8c39a720b8/pkg/storage/store.go#L91)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This constant isn't exported so it doesn't need a comment unless there's something more interesting to say. 
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 1413 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXp-CCr96A-xHF0S7WT:-KXq6jAtFYiPTSm5i7qT:b-y19wgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/eff3d2cbc3422b7410c547c58c35ff8c39a720b8/pkg/storage/store.go#L1413)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think the goroutine is needed to avoid deadlocks as this is called from contexts where `Store.mu` is held.
</blockquote></details>

It does need a new goroutine. I switched to using `stopper.RunAsyncTask`. It doesn't feel like it needs a limited version because we only send the gossip maximum once via this method for any single successful run through `Store.GossipStores`, which is where the countdown values are reset.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-:-KXq8K8BZfueYivASRBX:b-bovs04)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/storage/store.go, line 1421 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXqKqt-71p5mD9J9xzw:-KXqKqt-71p5mD9J9xzx:bwvotf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6489242f62efd53ba8cb94d5aaed0b1731463024/pkg/storage/store.go#L1421)):*
> ```Go
> 		atomic.StoreInt32(&s.gossipLeaseCountdown, 0)
> 		// Send using an async task because GossipStore needs the store mutex.
> 		ctx := s.AnnotateCtx(context.Background())
> ```

this should now be a plumbed context, since RunAsync takes care of giving you a safe derived context.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 5 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/storage/store.go, line 1421 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXqKqt-71p5mD9J9xzw:-KXqNAaXiazlkAEFZjdu:b5iyc12) ([raw file](https://github.com/cockroachdb/cockroach/blob/6489242f62efd53ba8cb94d5aaed0b1731463024/pkg/storage/store.go#L1421)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this should now be a plumbed context, since RunAsync takes care of giving you a safe derived context.
</blockquote></details>

plumbed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/storage/store.go, line 2158 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXqOwZImMotF0VDtQd9:-KXqOwZImMotF0VDtQdA:bb7ejjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/268a21bebda283c1d3490838e744cebe4efe485a/pkg/storage/store.go#L2158)):*
> ```Go
> 	// Add the range to metrics and maybe gossip on capacity change.
> 	s.metrics.ReplicaCount.Inc(1)
> 	s.maybeGossipOnCapacityChange(s.AnnotateCtx(context.Background()), rangeChangeEvent)
> ```

ctx?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks broke.

---

*[pkg/storage/store.go, line 2158 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662#-KXqOwZImMotF0VDtQd9:-KXqP_8XZHeMtdCqlXIK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/268a21bebda283c1d3490838e744cebe4efe485a/pkg/storage/store.go#L2158)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ctx?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11662)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> While this is all still fresh (and I have relatively high confidence besides), I'd like to see it deployed on its own so in the event it fails, I don't have to only find out about that sometime next week when we get around to deploying the removal of SnapshotResult.Capacity.

How are you going to detect if this mechanism fails? Curious to know what you're proactively planning to do here.

spencerkimball commented:
Well my thought was to keep an eye on the rebalancing graphs.

On Wed, Nov 30, 2016 at 2:07 PM Peter Mattis <notifications@github.com>
wrote:

> While this is all still fresh (and I have relatively high confidence
> besides), I'd like to see it deployed on its own so in the event it fails,
> I don't have to only find out about that sometime next week when we get
> around to deploying the removal of SnapshotResult.Capacity.
>
> How are you going to detect if this mechanism fails? Curious to know what
> you're proactively planning to do here.
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/11662#issuecomment-263964304>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF3MTaYJYt-wMNOmUb9EOYFAfmAWEAJ8ks5rDcl2gaJpZM4K99ST>
> .
>
